File: small_world.py
import time

class Frog:
    def __init__(self):
        self.hungry = True

    def day_of_frog(self):
        time.sleep(2)
        print("Frog:is breathing, moving")
        if self.hungry:
            time.sleep(2)
            print("Frog is eating grass")
        self.hungry = False

    def night_of_frog(self):
        time.sleep(2)
        print("Frog:is sleeping")
        self.hungry = True

class Tree:
    def __init__(self, frog: Frog):
        self.oxygen = False
        self.frog = frog

    def day_of_tree(self):
        self.oxygen = True
        time.sleep(2)
        print("Tree:is creating oxygen")
        self.frog.day_of_frog()

    def night_of_tree(self):
        self.oxygen = False
        time.sleep(2)
        print("Tree:is not creating oxygen")
        self.frog.night_of_frog()

class Sun:
    def __init__(self, light, tree: Tree):
        self.light = light
        self.tree = tree

    def day_in_world(self):
        while True:
            time.sleep(2)
            if self.light:
                print("Sun:is shining")
                self.tree.day_of_tree()
                self.light = False
            if not self.light:
                print("Sun:is not shining")
                self.tree.night_of_tree()
                self.light = True


frog = Frog()
tree = Tree(frog)
sun = Sun(True, tree)
sun.day_in_world()

File: small_world.c
#include <stdio.h>
#include <stdbool.h>
#include <unistd.h>

typedef struct Frog {
  bool hungry;
} Frog;

void day_of_frog(Frog* frog) {
  sleep(2);
  printf("Frog is breathing, moving\n");
  if (frog->hungry) {
    sleep(2);
    printf("Frog is eating grass\n");
  }
  frog->hungry = false;
}

void night_of_frog(Frog* frog) {
  sleep(2);
  printf("Frog is sleeping\n");
  frog->hungry = true;
}

typedef struct Tree {
  bool oxygen;
  Frog* frog;
} Tree;

void day_of_tree(Tree* tree) {
  tree->oxygen = true;
  sleep(2);
  printf("Tree is creating oxygen\n");
  day_of_frog(tree->frog);
}

void night_of_tree(Tree* tree) {
  tree->oxygen = false;
  sleep(2);
  printf("Tree is not creating oxygen\n");
  night_of_frog(tree->frog);
}

typedef struct Sun {
  bool light;
  Tree* tree;
} Sun;

void day(Sun* sun) {
  while (true) {
    sleep(2);
    if (sun->light) {
      printf("Sun is shining\n");
      day_of_tree(sun->tree);
      sun->light = false;
    }
    if (!sun->light) {
      printf("Sun is not shining\n");
      night_of_tree(sun->tree);
      sun->light = true;
    }
  }
}

int main(int argc, char** argv) {
  Frog frog = { true };
  Tree tree = { false, &frog };
  Sun sun = { true, &tree };
  day(&sun);
  return 0;
}
